File: tests/e2e/fixtures/code/sample.fs
// F#: records, discriminated unions, pipelines, pattern matching.
module Scanner

let limit = 5L * 1024L * 1024L

type Note = { Path: string; Tags: string list; Size: int64 }

type Mode =
    | Preview
    | Edit of force: bool

let isLarge (n: Note) = n.Size > limit

let groupByTag (notes: Note list) =
    notes
    |> List.filter (isLarge >> not)
    |> List.collect (fun n -> n.Tags |> List.map (fun t -> t, n))
    |> List.groupBy fst
    |> List.map (fun (t, ps) -> t, List.map snd ps)
    |> Map.ofList

let describe n =
    match n with
    | 0 -> "no tags"
    | n when n > 100 -> sprintf "many tags: %d" n
    | n -> sprintf "%d tags" n

printfn "%s" (describe (groupByTag [ { Path = "a.md"; Tags = [ "x" ]; Size = 12L } ] |> Map.count))

File: tests/e2e/fixtures/code/sample.ml
(* OCaml: records, variants, pattern matching, modules. *)
let limit = 5 * 1024 * 1024

type note = { path : string; tags : string list; size : int }

type mode = Preview | Edit of bool

let is_large n = n.size > limit

let group_by_tag notes =
  let tbl = Hashtbl.create 16 in
  List.iter
    (fun n ->
      if not (is_large n) then
        List.iter (fun t -> Hashtbl.replace tbl t (n :: (try Hashtbl.find tbl t with Not_found -> []))) n.tags)
    notes;
  tbl

let describe = function
  | 0 -> "no tags"
  | n when n > 100 -> Printf.sprintf "many tags: %d" n
  | n -> Printf.sprintf "%d tags" n

let () = print_endline (describe (Hashtbl.length (group_by_tag [ { path = "a.md"; tags = [ "x" ]; size = 12 } ])))
